File: IGListTest/IGListTest/Systems/Pathfinder.m
//
//  Pathfinder.m
//  IGListTest
//
//  Created by liuruixuan on 2023/4/16.
//

#import "Pathfinder.h"
#import "MarslinkModel.h"

@implementation Pathfinder

- (instancetype)init
{
    self = [super init];
    if (self) {
        [self addObserver:self forKeyPath:NSStringFromSelector(@selector(messages)) options:(NSKeyValueObservingOptionNew) context:nil];
    }
    return self;
}

- (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(NSDictionary<NSKeyValueChangeKey,id> *)change context:(void *)context {
    if ([keyPath isEqualToString:@"messages"]) {
        if (self.delegate && [self.delegate respondsToSelector:@selector(pathfinderDidUpdateMessagesWithPathfinder:)]) {
            [self.delegate pathfinderDidUpdateMessagesWithPathfinder:self];
        }
    }
}

- (void)connect {
    NSMutableArray *tempArray = [self mutableArrayValueForKey:NSStringFromSelector(@selector(messages))];
    [self delayWithTime:2.3 excute:^{
        [tempArray addObject:[self lewisMessageWithText:@"Liftoff in 3..." interval:0]];
        [self delayWithTime:1 excute:^{
            [tempArray addObject:[self lewisMessageWithText:@"2..." interval:0]];
            [self delayWithTime:1 excute:^{
                [tempArray addObject:[self lewisMessageWithText:@"1..." interval:0]];
            }];
        }];
    }];
}

- (void)delayWithTime:(double)time excute:(void(^)(void))work {
    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)((DISPATCH_TIME_NOW + time) * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
        work();
    });
}

- (Message *)lewisMessageWithText:(NSString *)text interval:(NSTimeInterval)interval {
    User *user = [[User alloc] initWithId:2 name:@"cpt.lewis"];
    return [[Message alloc] initWithDate:[NSDate dateWithTimeIntervalSinceNow:interval] text:text user:user];
}

- (NSMutableArray<Message *> *)messages {
    if (!_messages) {
        _messages = [NSMutableArray array];
        NSMutableArray *tempArray = [self mutableArrayValueForKey:NSStringFromSelector(@selector(messages))];
        [tempArray addObject:[self lewisMessageWithText:@"Mark, are you receiving me?" interval:-803200]];
        [tempArray addObject:[self lewisMessageWithText:@"I think I left behind some ABBA, might help with the drive 😜" interval:-259200]];
    }
    return _messages;
}

- (void)dealloc {
    [self removeObserver:self forKeyPath:@"messages"];
}

@end

File: Marslink_Swift5/Marslink_Swift5/Systems/Pathfinder.swift
//
//  Pathfinder.swift
//  Marslink_Swift5
//
//  Created by liuruixuan on 2023/4/17.
//

import Foundation

protocol PathfinderDelegate: AnyObject {
    func pathfinderDidUpdateMessages(pathfinder: Pathfinder)
}

private func delay(time: Double = 1, execute work: @escaping @convention(block) () -> Swift.Void) {
    DispatchQueue.main.asyncAfter(deadline: DispatchTime.now() + time) {
        work()
    }
}

private func lewisMessage(text: String, interval: TimeInterval = 0) -> Message {
    let user = User(id: 2, name: "cpt.lewis")
    return Message(date: Date(timeIntervalSinceNow: interval), text: text, user: user)
}

class Pathfinder {
    weak var delegate: PathfinderDelegate?

    var messages: [Message] = {
        var arr = [Message]()
        arr.append(lewisMessage(text: "Mark, are you receiving me?", interval: -803200))
        arr.append(lewisMessage(text: "I think I left behind some ABBA, might help with the drive 😜", interval: -259200))
        return arr
    }() {
        didSet {
            delegate?.pathfinderDidUpdateMessages(pathfinder: self)
        }
    }

    func connect() {
        delay(time: 2.3) {
            self.messages.append(lewisMessage(text: "Liftoff in 3..."))
            delay {
                self.messages.append(lewisMessage(text: "2..."))
                delay {
                    self.messages.append(lewisMessage(text: "1..."))
                }
            }
        }
    }
}
